File: src/record/TableScan.java
package record;

import file.BlockId;
import tx.Transaction;

public class TableScan {
    private Transaction tx;
    private Layout layout;
    private RecordPage rp;
    private String filename;
    private int currentslot;

    public TableScan(Transaction tx, String tblname, Layout layout) {
        this.tx = tx;
        this.layout = layout;
        filename = tblname + ".tbl";
        if (tx.size(filename) == 0) {
            moveToNewBlock();
        } else {
            moveToBlock(0);
        }
    }

    public void close() {
        if (rp != null) {
            tx.unpin(rp.block());
        }
    }

    public void beforeFirst() {
        moveToBlock(0);
    }

    public boolean next() {
        currentslot = rp.nextAfter(currentslot);
        while (currentslot < 0) {
            if (atLastBlock()) {
                return false;
            }
            moveToBlock(rp.block().number() + 1);
            currentslot = rp.nextAfter(currentslot);
        }
        return true;
    }

    public int getInt(String fldname) {
        return rp.getInt(currentslot, fldname);
    }

    public String getString(String fldname) {
        return rp.getString(currentslot, fldname);
    }

    public boolean hasField(String fldname) {
        return layout.schema().hasField(fldname);
    }

    public void setInt(String fldname, int val) {
        rp.setInt(currentslot, fldname, val);
    }

    public void setString(String fldname, String val) {
        rp.setString(currentslot, fldname, val);
    }

    public void insert() {
        currentslot = rp.insertAfter(currentslot);
        while (currentslot < 0) {
            if (atLastBlock()) {
                moveToNewBlock();
            } else {
                moveToBlock(rp.block().number() + 1);
            }
            currentslot = rp.insertAfter(currentslot);
        }
    }

    public void delete() {
        rp.delete(currentslot);
    }

    public void moveToRid(RID rid) {
        close();
        BlockId blk = new BlockId(filename, rid.blockNumber());
        rp = new RecordPage(tx, blk, layout);
        currentslot = rid.slot();
    }

    public RID getRid() {
        return new RID(rp.block().number(), currentslot);
    }

    private void moveToBlock(int blknum) {
        close();
        BlockId blk = new BlockId(filename, blknum);
        rp = new RecordPage(tx, blk, layout);
        currentslot = -1;
    }

    private void moveToNewBlock() {
        close();
        BlockId blk = tx.append(filename);
        rp = new RecordPage(tx, blk, layout);
        rp.format();
        currentslot = -1;
    }

    private boolean atLastBlock() {
        return rp.block().number() == tx.size(filename) - 1;
    }
}

File: src/main/kotlin/simpledb/record/rowlock/TableScan.kt
package simpledb.record.rowlock

import simpledb.file.BlockId
import simpledb.query.Constant
import simpledb.record.Layout
import simpledb.record.RID
import simpledb.tx.rowlock.TransactionImpl
import java.sql.Types

class TableScan(
    private val tx: TransactionImpl,
    tableName: String,
    private val layout: Layout
)  {
    private var rp: RecordPage? = null
    private val filename: String
    private var currentSlot = 0

    init {
        filename = "$tableName.tbl"
        if (tx.size(filename) == 0) {
            moveToNewBlock()
        } else {
            moveToBlock(0)
        }
    }

    fun close() {
        if (rp != null) {
            tx.unpin(rp!!.block())
        }
    }

    fun getVal(fieldName: String): Constant {
        return if (layout.schema()
            .type(fieldName) == Types.INTEGER
        ) {
            Constant(getInt(fieldName))
        } else {
            Constant(getString(fieldName))
        }
    }

    fun beforeFirst() {
        moveToBlock(0)
    }

    fun next(): Boolean {
        currentSlot = rp!!.nextAfter(currentSlot)
        while (currentSlot < 0) {
            if (atLastBlock()) {
                return false
            }
            moveToBlock(rp!!.block().number() + 1)
            currentSlot = rp!!.nextAfter(currentSlot)
        }
        return true
    }

    fun getInt(fldname: String): Int {
        return rp!!.getInt(currentSlot, fldname)
    }

    fun getString(fldname: String): String {
        return rp!!.getString(currentSlot, fldname)
    }

    fun hasField(fldname: String): Boolean {
        return layout.schema().hasField(fldname)
    }

    fun setVal(fieldName: String, value: Constant) {
        if (layout.schema().type(fieldName) == Types.INTEGER) {
            setInt(fieldName, value.asInt())
        } else {
            setString(
                fieldName,
                value.asString()
            )
        }
    }

    fun setInt(fldname: String, value: Int): Boolean =
        rp!!.setInt(currentSlot, fldname, value)

    fun setString(fldname: String, value: String): Boolean =
        rp!!.setString(currentSlot, fldname, value)

    fun insert() {
        currentSlot = rp!!.insertAfter(currentSlot)
        while (currentSlot < 0) {
            if (atLastBlock()) {
                moveToNewBlock()
            } else {
                moveToBlock(rp!!.block().number() + 1)
            }
            currentSlot = rp!!.insertAfter(currentSlot)
        }
    }

    fun delete() {
        rp!!.delete(currentSlot)
    }

    fun moveToRid(rid: RID) {
        close()
        val blk = BlockId(filename, rid.blockNumber())
        rp = RecordPage(tx, blk, layout)
        currentSlot = rid.slot()
    }

    fun getRid(): RID {
        return RID(rp!!.block().number(), currentSlot)
    }

    private fun moveToBlock(blkNum: Int) {
        close()
        val blk = BlockId(filename, blkNum)
        rp = RecordPage(tx, blk, layout)
        currentSlot = -1
    }

    private fun moveToNewBlock() {
        close()
        val blk = tx.append(filename)
        rp = RecordPage(tx, blk, layout)
        rp!!.format()
        currentSlot = -1
    }

    private fun atLastBlock(): Boolean {
        return rp!!.block().number() == tx.size(filename) - 1
    }
}
